### ISSUE TITLE
[Feature] Support `ConcretizedCallable` as parameter type for parametric gates with adjoint differentiation mode 

### ISSUE THREAD

**Issue by vytautas-a** [pasqal-io/pyqtorch#284]
Support `adjoint` differentiation mode for parametric gates where the differentiable parameter is given as `ConcretizedCallable`.

**Comment by mlahariya** [comment]
@vytautas-a, Can you provide a bit more description on this?

Also, what could be a possible solution - should we keep the adjoint disabled for ConcretizedCallable? implement it seperately?

**Comment by RolandMacDoland** [comment]
@vytautas-a Is it still relevant ? I had the feeling we weren't using `ConcretizedCallable` anymore.

**Comment by chMoussa** [comment]
> [@vytautas-a](https://github.com/vytautas-a) Is it still relevant ? I had the feeling we weren't using `ConcretizedCallable` anymore.

@RolandMacDoland Yes we are still using it. So we can still consider it relevant.

**Comment by chMoussa** [comment]
> [@vytautas-a](https://github.com/vytautas-a), Can you provide a bit more description on this?
> 
> Also, what could be a possible solution - should we keep the adjoint disabled for ConcretizedCallable? implement it seperately?

Hey @mlahariya, no we should extend the adjoint with `ConcretizedCallable` support. I added extra context to @vytautas-a 's issue.

**Comment by Erik-Chagas** [comment]
Working on it!

**Comment by Erik-Chagas** [comment]
Hi @chMoussa @vytautas-a, do we need to add a new test in ``tests/differentiation/test_differentiation.py`` or change an existing test?

**Comment by chMoussa** [comment]
> Hi @chMoussa @vytautas-a, do we need to add a new test in ``tests/differentiation/test_differentiation.py`` or change an existing test?

Probably a new one.

**Comment by Erik-Chagas** [comment]
I have a question. Would the main task be to implement embedding for diff mode adjoint?

**Comment by chMoussa** [comment]
> I have a question. Would the main task be to implement embedding for diff mode adjoint?

@Erik-Chagas The task would be to indeed extend diff mode adjoint with embedding and adding tests for testing it out against the `AD` diff mode.

**Comment by chMoussa** [comment]
> @chMoussa, does this feature need to be supported without using Embedding?

Hello @neogyk , I think it mostly needs the ConcretizedCallable. What would be an example without Embedding in this case?

**Comment by chMoussa** [comment]
> [@chMoussa](https://github.com/chMoussa). The case without embedding will be:
> 
> ```
> import torch
> import pyqtorch as pyq
> from pyqtorch.utils import DiffMode
> from pyqtorch.embed import ConcretizedCallable
> 
> 
> x = torch.rand(1, requires_grad=True)
> 
> sin_x_fn = pyq.ConcretizedCallable(call_name = 'sin', abstract_args= ['x'])
> circ = pyq.QuantumCircuit(1, [pyq.RX(0, sin_x_fn)])
> state= pyq.zero_state(1)
> y = torch.rand(1, requires_grad=True)
> values = {'x': x}
> obs = pyq.Observable([pyq.Z(0)])
> expval = pyq.expectation(circuit=circ, state=state, values=values, observable=obs, diff_mode=pyq.DiffMode.ADJOINT)
> print("Expval:", expval)
> print(torch.autograd.grad(expval, (x, y), torch.ones_like(expval),allow_unused=True))
> ```

@neogyk Yes that looks good to me.

**Comment by chMoussa** [comment]
@Erik-Chagas @neogyk FYI the deadline for unitary hack is coming soon, I suggest to open a PR with what you have, maybe even collaborate and share.
